src/blob_store.rs

```diff
@@ -0,0 +1,260 @@
+//! The `blob_store` module provides functions for parallel verification of the
+//! Proof of History ledger as well as iterative read, append write, and random
+//! access read to a persistent file-based ledger.
+// TODO: Remove when implementation complete
+// TODO: use parallelism at slots? should be able to max disk eventually...
+// probably set up some kind of write queue that can accept 'jobs' via a channel or something
+// and then farm
+#![allow(unreachable_code, unused_imports, unused_variables, dead_code)]
+
+use crate::db_ledger::SlotMeta;
+use crate::entry::Entry;
+use crate::packet::Blob;
+
+use serde::Serialize;
+
+use std::borrow::Borrow;
+use std::error::Error;
+use std::fmt;
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, prelude::*, BufReader, Seek, SeekFrom};
+use std::path::{Path, PathBuf};
+use std::result::Result as StdRes;
+
+mod store_impl;
+#[cfg(test)]
+mod tests;
+
+type Result<T> = StdRes<T, StoreError>;
+
+#[derive(Debug)]
+pub struct Store {
+    root: PathBuf,
+}
+
+#[derive(Debug)]
+pub enum StoreError {
+    Io(io::Error),
+    BadBlob,
+    NoSuchBlob(u64, u64),
+    NoSuchSlot(u64),
+    Bincode(bincode::Error),
+}
+
+#[derive(Copy, Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+struct BlobIndex {
+    pub index: u64,
```

Comment:
is this implicit in the location in the index file, i.e. do you need the number in the record?

Reply:
it's not implicit in the location because the blobs can get stored out-of-order or with gaps. (unless there's some guarantee that won't happen?)

Reply:
doesn't matter if they come in out of order, since the index record is of fixed size, seek to the index offset and write the data file offset.

blobs are out of order in the data file, but information about where they landed is in order in the index file